### PR TITLE
UCT/IB: Upstream DM

### PR DESCRIPTION
**Pull request by Artemy-Mellanox**

### CODE REVIEW

File: src/uct/ib/configure.m4 [hunk 1]
```diff
@@ -361,7 +361,18 @@ AS_IF([test "x$with_ib" == xyes],
...
                ],
-               [AC_DEFINE([HAVE_IBV_EXP_DM], 1, [Device Memory support])],
+               [AC_DEFINE([HAVE_IBV_DM], 1, [Device Memory support])
+                AC_DEFINE([HAVE_IBV_EXP_DM], 1, [Device Memory support (EXP)])],
+               [])
+            AC_TRY_COMPILE([#include <infiniband/verbs.h>],
+               [
+                   struct ibv_dm ibv_dm;
+                   struct ibv_alloc_dm_attr dm_attr;
+                   void* a1 = ibv_alloc_dm;
```
Comment (yosefe) [inline src/uct/ib/configure.m4]: why not AC_CHECK_FUNCS + AC_CHECK_DECLS ?

File: src/uct/ib/mlx5/ib_mlx5_dv.h [hunk 2]
```diff
@@ -58,4 +64,51 @@ struct ibv_qp *uct_dv_get_cmd_qp(struct ibv_srq *srq);
...
+} uct_mlx5_dm_va_t;
+
+static ucs_status_t UCS_F_MAYBE_UNUSED
+uct_ib_mlx5_get_dm_info(struct ibv_exp_dm *dm, struct mlx5dv_dm *dm_info)
+{
+    dm_info->buf = ((uct_mlx5_dm_va_t*)dm)->start_va;
+    return UCS_OK;
+}
+#else
+# define dv_dm dv.dm
```
Comment (yosefe) [inline src/uct/ib/mlx5/ib_mlx5_dv.h]: too generic define

File: src/uct/ib/rc/accel/rc_mlx5_common.c [hunk 3]
```diff
@@ -447,28 +437,32 @@ uct_rc_mlx5_iface_common_dm_tl_init(uct_mlx5_dm_data_t *data,
...
+                                       IBV_ACCESS_REMOTE_ATOMIC);
     if (data->mr == NULL) {
-        ucs_warn("ibv_exp_reg_mr() error - On Device Memory registration failed, %d %m", errno);
+        ucs_warn("ibv_reg_mr_dm() error - On Device Memory registration failed, %d %m", errno);
         status = UCS_ERR_NO_RESOURCE;
         goto failed_mr;
     }
 
-    data->start_va = ((uct_mlx5_dm_va_t*)data->dm)->start_va;
+    obj.dv_dm.in = data->dm;
```
Comment (yosefe) [inline src/uct/ib/rc/accel/rc_mlx5_common.c]: align = on column

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/8882/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/8885/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/8886/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/8906/ for details (Mellanox internal link).

File: src/uct/ib/mlx5/ib_mlx5_dv.h [hunk 4]
```diff
@@ -107,8 +107,10 @@ uct_ib_mlx5_get_dm_info(struct ibv_exp_dm *dm, struct mlx5dv_dm *dm_info)
     dm_info->buf = ((uct_mlx5_dm_va_t*)dm)->start_va;
     return UCS_OK;
 }
+
+# define UCT_IB_MLX5_DV_DM(obj) obj.dv_dm
 #else
-# define dv_dm dv.dm
+# define UCT_IB_MLX5_DV_DM(obj) obj.dv.dm
```
Comment (yosefe) [inline src/uct/ib/mlx5/ib_mlx5_dv.h]: _obj

File: src/uct/ib/mlx5/ib_mlx5_dv.c [hunk 5]
```diff
@@ -20,6 +20,11 @@ ucs_status_t uct_ib_mlx5dv_init_obj(uct_ib_mlx5dv_t *obj, uint64_t type)
     int ret;
 
     ret = mlx5dv_init_obj(&obj->dv, type);
+#if HAVE_IBV_EXP_DM
```
Comment (yosefe) [inline src/uct/ib/mlx5/ib_mlx5_dv.c]: should it be HAVE_IBV_DM? since it uses MLX5DV_OBJ_DM?

File: src/uct/ib/mlx5/ib_mlx5_hw.c [hunk 6]
```diff
@@ -173,6 +173,12 @@ ucs_status_t uct_ib_mlx5dv_init_obj(uct_ib_mlx5dv_t *obj, uint64_t obj_type)
                 ucs_container_of(obj->dv.srq.out, uct_ib_mlx5dv_srq_t, dv));
     }
 
+#if HAVE_IBV_EXP_DM
```
Comment (yosefe) [inline src/uct/ib/mlx5/ib_mlx5_hw.c]: should it be HAVE_IBV_DM? since it uses MLX5DV_OBJ_DM?

File: src/uct/ib/mlx5/ib_mlx5_dv.c (hunk 5 above)
Comment (Artemy-Mellanox) [reply src/uct/ib/mlx5/ib_mlx5_dv.c]: It's fallback when there are no DM in mlx5dv

File: src/uct/ib/mlx5/ib_mlx5_hw.c (hunk 6 above)
Comment (Artemy-Mellanox) [reply src/uct/ib/mlx5/ib_mlx5_hw.c]: It's fallback when there are no mlx5dv

Comment (yosefe) [reply src/uct/ib/mlx5/ib_mlx5_hw.c]: so why need `#if`? maybe it should be inside uct_ib_mlx5_get_dm_info?

Comment (Artemy-Mellanox) [reply src/uct/ib/mlx5/ib_mlx5_hw.c]: if not defined HAVE_IBV_EXP_DM neither is MLX5DV_OBJ_DM - won't compile

Comment (yosefe) [reply src/uct/ib/mlx5/ib_mlx5_hw.c]: ok

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/8962/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/9414/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/9451/ for details (Mellanox internal link).

**Comment by Artemy-Mellanox** [comment]
bot:mlx:retest

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/9541/ for details (Mellanox internal link).

**Comment by Artemy-Mellanox** [comment]
bot:bgate:retest

**Pull request merged**